### PR TITLE
Tests for RawSwarm

### PR DESCRIPTION
**Pull request by dvdplm**
Based on https://github.com/libp2p/rust-libp2p/pull/588.

### CODE REVIEW

File: core/src/tests/dummy_transport.rs [hunk 1]
```diff
@@ -85,14 +91,31 @@ impl Transport for DummyTransport {
...
+                future::ok((peer_id, DummyMuxer::new()))
+            };
+        Ok(Box::new(fut))
     }
 
-    fn nat_traversal(&self, _server: &Multiaddr, _observed: &Multiaddr) -> Option<Multiaddr> {
-        unimplemented!();
+    fn nat_traversal(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        if server == observed {
+            Some(observed.encapsulate("/udt").unwrap())
```
Comment (tomaka) [inline core/src/tests/dummy_transport.rs]: I know this is a dummy, but this change doesn't really make sense.

File: core/src/tests/dummy_transport.rs [hunk 2]
```diff
@@ -85,14 +91,31 @@ impl Transport for DummyTransport {
...
     {
-        unimplemented!();
+        let peer_id = if let Some(peer_id) = self.next_peer_id {
+            peer_id
+        } else {
+            PublicKey::Rsa((0 .. 128).map(|_| -> u8 { 1 }).collect()).into_peer_id()
+        };
+
+        let fut =
+            if addr.to_string() == "/unix/unreachable" {
```
Comment (tomaka) [inline core/src/tests/dummy_transport.rs]: Same here. This is extremely arbitrary.

File: core/src/tests/dummy_transport.rs [hunk 3]
```diff
@@ -85,14 +91,31 @@ impl Transport for DummyTransport {
...
-    fn dial(self, _addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)>
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)>
     where
         Self: Sized,
     {
-        unimplemented!();
+        let peer_id = if let Some(peer_id) = self.next_peer_id {
+            peer_id
+        } else {
+            PublicKey::Rsa((0 .. 128).map(|_| -> u8 { 1 }).collect()).into_peer_id()
```
Comment (tomaka) [inline core/src/tests/dummy_transport.rs]: Should be `PeerId::random()`.

File: core/src/nodes/listeners.rs [hunk 4]
```diff
@@ -504,7 +518,9 @@ mod tests {
     #[test]
     fn listener_stream_poll_processes_listeners_in_turn() {
         let mut t = DummyTransport::new();
-        t.set_initial_listener_state(ListenerState::Ok(Async::Ready(Some(1))));
+        let peer_id = PublicKey::Rsa((0 .. 128).map(|_| -> u8 { 1 }).collect()).into_peer_id();
```
Comment (tomaka) [inline core/src/nodes/listeners.rs]: Should be `PeerId::random()`.

File: core/src/nodes/listeners.rs [hunk 5]
```diff
@@ -441,7 +446,9 @@ mod tests {
     #[test]
     fn listener_stream_poll_with_erroring_listener_emits_closed_event() {
         let mut t = DummyTransport::new();
-        t.set_initial_listener_state(ListenerState::Ok(Async::Ready(Some(1))));
+        let peer_id = PublicKey::Rsa((0 .. 128).map(|_| -> u8 { 1 }).collect()).into_peer_id();
```
Comment (tomaka) [inline core/src/nodes/listeners.rs]: Should be `PeerId::random()`.

File: core/src/nodes/listeners.rs [hunk 6]
```diff
@@ -455,8 +462,10 @@ mod tests {
     #[test]
     fn listener_stream_poll_chatty_listeners_each_get_their_turn() {
         let mut t = DummyTransport::new();
-        t.set_initial_listener_state(ListenerState::Ok(Async::Ready(Some(1))));
-        let mut ls = ListenersStream::new(t);
+        let peer_id = PublicKey::Rsa((0 .. 128).map(|_| -> u8 { 1 }).collect()).into_peer_id();
```
Comment (tomaka) [inline core/src/nodes/listeners.rs]: Should be `PeerId::random()`.

File: core/src/nodes/listeners.rs [hunk 7]
```diff
@@ -382,10 +385,13 @@ mod tests {
     }
 
     #[test]
-    fn listener_stream_poll_with_ready_listeners_yields_upgrade() {
-        let mut transport = DummyTransport::new();
-        transport.set_initial_listener_state(ListenerState::Ok(Async::Ready(Some(1))));
-        let mut ls = ListenersStream::new(transport);
+    fn listener_stream_poll_with_ready_listeners_is_ready() {
+        let mut t = DummyTransport::new();
+        let peer_id = PublicKey::Rsa((0 .. 128).map(|_| -> u8 { 1 }).collect()).into_peer_id();
```
Comment (tomaka) [inline core/src/nodes/listeners.rs]: Should be `PeerId::random()`.

File: core/src/nodes/raw_swarm.rs [hunk 8]
```diff
@@ -192,6 +195,83 @@ where
     },
 }
 
+impl<'a, TTrans, TInEvent, TOutEvent, THandler> fmt::Debug for RawSwarmEvent<'a, TTrans, TInEvent, TOutEvent, THandler>
```
Comment (tomaka) [inline core/src/nodes/raw_swarm.rs]: This should have been in a different pull request, but let's not remove it now.

Comment (dvdplm) [reply core/src/nodes/raw_swarm.rs]: Reason was I needed a `Debug` impl in order to make asserts and I didn't want to block this PR needlessly.

File: core/src/tests/dummy_transport.rs (hunk 1 above)
Comment (dvdplm) [reply core/src/tests/dummy_transport.rs]: Yeah, it smells for sure. So the context is [this test](https://github.com/libp2p/rust-libp2p/pull/602/files#diff-0c5110fdcca8841806bc220275bed002R1268) where I test that "nat_traversal_transforms_the_observed_address_according_to_the_transport_used", so I need it to perform some kind of transform to assert on. While I agree that tossing on `/udt` is silly, I'm not sure what kind of transform a `DummyTransport` should be performing. I could just toss back the `observed` but that's also a little unsatisfying.
Do you have an idea for what is better here? pop off the port and, dunno, increment it by one?

File: core/src/tests/dummy_transport.rs (hunk 2 above)
Comment (dvdplm) [reply core/src/tests/dummy_transport.rs]: Here the context are these two tests: [one](https://github.com/libp2p/rust-libp2p/pull/602/files#diff-0c5110fdcca8841806bc220275bed002R1470), [two](https://github.com/libp2p/rust-libp2p/pull/602/files#diff-0c5110fdcca8841806bc220275bed002R1498) that tests that the appropriate errors are propagated. I needed a way to emulate an unreachable host.
What's a better way? I could perhaps make an "EvenDummerDummyTransport" with this hack in it. Or add a test-only "unreachable" protocol variant?

Comment (dvdplm) [reply core/src/tests/dummy_transport.rs]: I changed this to use the IPv6 "black hole" prefix, `0100/::64`. Not sure if that's any better though. :/

File: core/src/tests/dummy_transport.rs (hunk 1 above)
Comment (dvdplm) [reply core/src/tests/dummy_transport.rs]: I changed this to act more like other `nat_traversal` impls and made it increment the port by one. Maybe slightly less nonsensical. :/

**Pull request merged**